rke_tests/common.py

```diff
@@ -0,0 +1,214 @@
+        dns service discovery validation
+    """
+
+    node_name = '{}-node'.format(test_name)
+
+    # create node from cloud provider
+    nodes = cloud_provider.create_multiple_nodes(
+        number_of_nodes, node_name, wait_for_ready=True)
+
+    try:
+        # create rke cluster yml
+        config_yml, nodes = rke_client.build_rke_template(rke_template, nodes)
+
+        # run rke up
+        result = rke_client.up(config_yml)
+        assert result.ok, result.stderr
+
+        # validate k8s reachable
+        kubectl.kube_config_path = rke_client.kube_config_path()
+        validation_node_roles(nodes, kubectl.get_nodes())
+        validate_private_ip_intercommunication(kubectl)
+        validate_dns_service_discovery(kubectl)
+    finally:
+        for node in nodes:
+            cloud_provider.delete_node(node)
+
+
+def match_nodes(nodes, k8s_nodes):
+    """
+    Builds a list of tuples, where:
+    nodes_to_k8s_nodes[0][0] is the node object matched to
+    nodes_to_k8s_nodes[0][1] is the k8s info for the same node
+    TODO: what about etcd only nodes? are they returned in kubectl get nodes?
+    """
+
+    nodes_to_k8s_nodes = []
+    for k8s_node in k8s_nodes['items']:
+        hostname = k8s_node['metadata']['labels']['kubernetes.io/hostname']
+        for node in nodes:
+            if hostname == node.public_ip_address or \
+                    hostname == node.host_name:
+                nodes_to_k8s_nodes.append((node, k8s_node))
+    return nodes_to_k8s_nodes
+
+
+def assert_containers_exist_for_role(role, containers, expect_containers):
+    missing_containers = expect_containers[:]
+    for container in containers:
+        if container in expect_containers:
+            missing_containers.remove(container)
+    assert len(missing_containers) == 0, \
+        "Missing expected containers for role '{0}': {1}".format(
+            role, missing_containers)
+
+
+def validation_node_roles(nodes, k8s_nodes):
+    """
+    Validates each node's labels for match its roles
+    Validates each node's running containers match its role
+    Validates etcd etcdctl cluster-health command
+    Validates worker nodes nginx-proxy conf file for controlplane ips
+    """
+    # TODO: etcd only nodes are not listed in k8s_nodes, should handle this
+    # TODO: does hostname_override affect nginx-proxy controlplane ips?
+
+    role_matcher = {
+        'worker': 'node-role.kubernetes.io/worker',
+        'etcd': 'node-role.kubernetes.io/etcd',
+        'controlplane': 'node-role.kubernetes.io/master'}
+    controlplane = [
+        'scheduler', 'kube-controller', 'kubelet', 'kube-proxy', 'kube-api']
+    worker = ['kubelet', 'kube-proxy', 'nginx-proxy']
+    etcd = ['etcd']
+
+    controlplane_ips = []
+    for node in nodes:
+        if 'controlplane' in node.roles:
+            controlplane_ips.append(node.public_ip_address)
+
+    nodes_to_k8s_nodes = match_nodes(nodes, k8s_nodes)
+    for node, k8s_node in nodes_to_k8s_nodes:
```

Comment:
Change this logic to go over all nodes from the nodes list instead of nodes fetched from kubectl command